kuzzleio/kuzzle
# Implement a dedicated User Controller

## What does this PR do?

Towards the goal of splitting our huge `SecurityController`, creates a new standalone `UserController`, adds an abstraction layer `NativeSecurityController` and updates the `Security Controller`.

### Standalone UserController

1. Update documentation, deprecate old actions and migrate doc for new actions
2. Migrate and rename actions from SecurityController to UserController, deprecate the old ones
3. Update HTTP routes for each actions
4. Migrate unit tests for the new controller and check that new actions are called by old controller
5. Migrate or write functional tests for migrated actions

#### Summary of migrations

| SecurityController | UserController | Special Case |
| ------------------------ | ------------------- | ------------------ |
| `createUser` | `create` |
| `createRestrictedUser` | | Migrated in the AuthController as `auth:signin` |
| `getUser`| `get`|
| `mGetUsers`| `mGet`|
| `searchUsers`| `search`|
| `scrollUsers`| `scroll`|
| `updateUser`| `update`|
| `replaceUser`| `replace`|
| `deleteUser`| `delete`|
| `mDeleteUsers`| `mDelete`|
| `getUserMapping`| `getMappings`| Result: `{ mapping: {...} }` => `{ mappings, dynamic, _meta }` |
| `updateUserMapping`| `updateMappings`| Result: `{ properties: {...} }` => `{ properties, dynamic, _meta }` |
| `getUserRights`| `getRights`| Result: `{ hits: [...] }` => `{ rights: [...] }` |
| `checkRights`| `isAllowed`|
| `getUserStrategies`| `getStrategies`| Result: `{ strategies: [...], total: x }` => `{ strategies: [...] }` |
| `revokeTokens`| `revokeTokens`|
| `refresh` | `refresh`| `security:refresh` must still be used for `role` or `profile` collection |

#### About permissions

It would be costly, fastidious and useless after Kuzzle v3 to keep the same permissions between `security:oldAction` and `user:newAction`. Since Kuzzle determines permissions on a whitelist basis, every newly created actions will not be allowed unless it is explicitely stated. In that case, the developer is obviously up to date and **should be aware of this duplication of actions**. It is now his job to **watch out for incoherent permissions** during transition.

### AbstractSecurityController

From now, a security-related controller (aka `UserController`, `SecurityController`...) will extend `AbstractSecurityController` to take advantage of private generic functions since they share common logic.

#### Inheritance

- `UsualController` > `NativeController` > `BaseController`
- `SecurityRelatedController` > `AbstractSecurityController` > `NativeController` > `BaseController`

#### Generic methods

| Generic name | Used by |
| ------------------- | ------------- |
| `_get` | `security:getProfile`, `security:getRole`, `user:get` |
| `_mGet` | `security:mGetProfiles`, `security:mGetRoles`, `user:mGet` |
| `_delete` | `security:deleteProfile`, `security:deleteRole`, `user:delete` |
| `_mDelete` | `security:mDeleteProfiles`, `security:mDeleteRoles`, `user:mDelete` |
| `_refresh` | `security:refresh`, `user:refresh` |

### Other changes

- References to user related routes in the documentation updated
- Merge SecurityController method `_getSearchPageSize` into `request.getSearchParams` allowing to easily set a maximum default size as well as a useful 'NotExceedMaxFetch' check
- `security:createFirstAdmin` has been deprecated in favor of `user:create` used with admin profile
- Private method `securityController._persist` used by `create*` actions has been merged with `'core:security:user:create'` which was only asked by the first one anyway
- Deprecation warnings during functional tests disabled

### How this should be manually tested?

1. Clone and run Kuzzle
2. Try an old action (`security:getUser` for example), you should see some deprecration message in the response
3. Try a new action (`user:get` for example), you can now use any user actions, go check the doc!

### What's next?
The next step to successfully split the `SecurityController` is to implement a [RoleController](https://github.com/kuzzleio/kuzzle/issues/1973).
While doing this, you may add new methods to the abstraction layer since 'roles' and 'profiles' share even more common logic (`_update` and `_search` might be some good candidates)

Closes #1972
